rust/feature-flags/src/cohorts/cohort_operations.rs

```diff
@@ -0,0 +1,684 @@
+    ///           }
+    ///         ]
+    ///       }
+    ///     ]
+    ///   }
+    /// }
+    /// ```
+    fn traverse_filters(
+        inner: &InnerCohortProperty,
+        dependencies: &mut HashSet<CohortId>,
+    ) -> Result<(), FlagError> {
+        for cohort_values in &inner.values {
+            for filter in &cohort_values.values {
+                if filter.is_cohort() {
+                    // Assuming the value is a single integer CohortId
+                    if let Some(cohort_id) = filter.value.as_i64() {
+                        dependencies.insert(cohort_id as CohortId);
+                    } else {
+                        return Err(FlagError::CohortFiltersParsingError);
+                    }
+                }
+                // NB: we don't support nested cohort properties, so we don't need to traverse further
+            }
+        }
+        Ok(())
+    }
+}
+
+impl InnerCohortProperty {
+    /// Flattens the nested cohort property structure into a list of property filters.
+    ///
+    /// The cohort property structure in Postgres looks like:
+    /// ```json
+    /// {
+    ///   "type": "OR",
+    ///   "values": [
+    ///     {
+    ///       "type": "OR",
+    ///       "values": [
+    ///         {
+    ///           "key": "email",
+    ///           "value": "@posthog.com",
+    ///           "type": "person",
+    ///           "operator": "icontains"
+    ///         },
+    ///         {
+    ///           "key": "age",
+    ///           "value": 25,
+    ///           "type": "person",
+    ///           "operator": "gt"
+    ///         }
+    ///       ]
+    ///     }
+    ///   ]
+    /// }
+    /// ```
+    pub fn to_inner(self) -> Vec<PropertyFilter> {
+        self.values
+            .into_iter()
+            .flat_map(|value| value.values)
+            .collect()
+    }
+
+    /// Evaluates a cohort property based on its type (AND/OR) and values.
+    ///
+    /// This function recursively evaluates the cohort property tree structure, handling both
+    /// property matches and nested cohort membership checks.
+    pub fn evaluate(
+        &self,
+        target_properties: &HashMap<String, Value>,
+        cohort_matches: &HashMap<CohortId, bool>,
+    ) -> Result<bool, FlagError> {
+        match self.prop_type {
+            CohortPropertyType::OR => {
+                for cohort_values in &self.values {
+                    if evaluate_cohort_values(cohort_values, target_properties, cohort_matches)? {
+                        return Ok(true);
+                    }
+                }
+                Ok(false)
+            }
+            CohortPropertyType::AND => {
+                for cohort_values in &self.values {
+                    if !evaluate_cohort_values(cohort_values, target_properties, cohort_matches)? {
+                        return Ok(false);
+                    }
+                }
+                Ok(true)
+            }
+        }
+    }
+}
```

Comment:
This is the big change for this PR that fixed the bug – it got kinda lost in the refactor, but here's really what I did:

For the `InnerCohortProperty` implementation, I added methods to flatten nested structures and provided evaluation logic that respects the OR/AND tree structure

Then, for the `evaluate_cohort_values`, I did a better job of handling evaluation of cohort values based on different property types, and properly distinguishes between regular property matching and cohort membership checks.